Problem during concurrent tests

I got at least in one case (ubuntu, py3.5) an error on the concurrent tests.
The error was from one of the workers, that raised this error: 
https://github.com/giovannipizzi/disk-objectstore/blob/af8e70f2c168f4df3ec3fdf70ffeb9609577bf8c/disk_objectstore/utils.py#L304-L307

If this appears again, we need to better benchmark it (possibly printing the output of *all* concurrent jobs, so we can check what happened at the same time.

In a recent test (with a slightly changed output), I get
```
AssertionError: Mismatch for d7f69ad18a481e84250ef61e8bcbd9817914b98f45bed244b2e68f1f7a9c95ce: d7f69ad18a481e84250ef61e8bcbd9817914b98f45bed244b2e68f1f7a9c95ce vs. e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Note that `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855` is the sha256 of an empty string.
And if I don't delete the repository, and check the object with hash key `d7f69ad18a481e84250ef61e8bcbd9817914b98f45bed244b2e68f1f7a9c95ce`, it's there and it's not corrupted.

So I think what is happening is that the library returns an empty string when it shouldn't. Possible cause: it starts reading a loose object, and while it's being read, it also gets packed and deleted, and it returns an empty string? To be further investigated.

--- a/disk_objectstore/container.py
+++ b/disk_objectstore/container.py
@@ -15,7 +15,9 @@ from sqlalchemy.sql import func
 from sqlalchemy.orm import sessionmaker
 
 from .models import Base, Obj
-from .utils import (nullcontext, HashWriterWrapper, ObjectWriter, PackedObjectReader, StreamDecompresser, is_known_hash)
+from .utils import (
+    HashWriterWrapper, ObjectWriter, PackedObjectReader, StreamDecompresser, chunk_iterator, is_known_hash, nullcontext
+)
 from .exceptions import NotExistent, NotInitialised
 
 ObjQueryResults = namedtuple('ObjQueryResults', ['hashkey', 'offset', 'length', 'compressed', 'size'])
@@ -31,6 +33,18 @@ class Container:
     # Size in bytes of each of the chunks used when (internally) reading or writing in chunks, e.g.
     # when packing.
     _CHUNKSIZE = 65536
+
+    # When performing an `in_` query in SQLite, this is converted to something like
+    # 'SELECT * FROM db_object WHERE db_object.hashkey IN (?, ?)' with parameters = ('hash1', 'hash2')
+    # Now, the maximum number of parameters is limited in SQLite, see variable SQLITE_MAX_VARIABLE_NUMBER
+    # as described in https://www.sqlite.org/limits.html
+    # Now, until recently (at the moment of writing) this defaults to 999 for SQLite versions
+    # prior to 3.32.0 (2020-05-22) or 32766 for SQLite versions after 3.32.0.
+    # So we need to assume that we cannot put more than 999 elements in the `.in_` parameter.
+    # Note that on some OSs, the value is increased at compile time. E.g. on my Mac OS X with python 3.6
+    # compiled with HomeBrew, the limit (I tested it) is 250000.
+    # See also e.g. this comment https://bugzilla.redhat.com/show_bug.cgi?id=1798134
+    _IN_SQL_MAX_LENGTH = 950
 
     def __init__(self, folder):
         """Create the class that represents the container.
@@ -407,6 +421,9 @@ class Container:
             # The try/finally block makes sure we close it at the end, if any was open.
             last_open_file = None
 
+            # Operate on a set - only return once per required hashkey, even if required more than once
+            hashkeys_set = set(hashkeys)
+
             try:
                 hashkeys_in_packs = set()
 
@@ -415,12 +432,16 @@ class Container:
                 # a problem as we then split them by pack). To be checked, performance-wise, if it's better
                 # to order in python instead
                 session = self._get_cached_session()
-                query = session.query(Obj).filter(
-                    Obj.hashkey.in_(hashkeys)
-                ).with_entities(Obj.pack_id, Obj.hashkey, Obj.offset, Obj.length, Obj.compressed,
-                                Obj.size).order_by(Obj.offset)
-                for res in query:
-                    packs[res[0]].append(ObjQueryResults(res[1], res[2], res[3], res[4], res[5]))
+
+                # Operate in chunks, due to the SQLite limits
+                # (see comment above the definition of self._IN_SQL_MAX_LENGTH)
+                for chunk in chunk_iterator(hashkeys_set, size=self._IN_SQL_MAX_LENGTH):
+                    query = session.query(Obj).filter(
+                        Obj.hashkey.in_(chunk)
+                    ).with_entities(Obj.pack_id, Obj.hashkey, Obj.offset, Obj.length, Obj.compressed,
+                                    Obj.size).order_by(Obj.offset)
+                    for res in query:
+                        packs[res[0]].append(ObjQueryResults(res[1], res[2], res[3], res[4], res[5]))
 
                 for pack_int_id, pack_metadata in packs.items():
                     hashkeys_in_packs.update(obj.hashkey for obj in pack_metadata)
@@ -441,14 +462,14 @@ class Container:
                 # Collect loose hash keys that are not found
                 # Reason: a concurrent process might have packed them,
                 # in the meantime.
-                loose_not_found = []
-                for loose_hashkey in set(hashkeys).difference(hashkeys_in_packs):
+                loose_not_found = set()
+                for loose_hashkey in hashkeys_set.difference(hashkeys_in_packs):
                     obj_path = self._get_loose_path_from_hashkey(hashkey=loose_hashkey)
                     try:
                         last_open_file = open(obj_path, mode='rb')
                         yield loose_hashkey, last_open_file, os.path.getsize(obj_path)
                     except FileNotFoundError:
-                        loose_not_found.append(loose_hashkey)
+                        loose_not_found.add(loose_hashkey)
                         continue
                     if last_open_file is not None:
                         if not last_open_file.closed:
@@ -460,11 +481,6 @@ class Container:
                 # are by now in the pack.
                 # If they are not, the object does not exist.
                 if loose_not_found:
-                    # In this case, I do *not* sort them - I expect
-                    # this situation to happen quite rarely, only while
-                    # packing, and it's ok to lose a bit of performance
-                    # in this rare case, but keep the code simpler.
-
                     # IMPORTANT. I need to close the session (and flush the
                     # self._session cache) to refresh the DB, otherwise since I am
                     # reading in WAL mode, I will be keeping to read from the "old"
@@ -478,17 +494,20 @@ class Container:
 
                     packs = defaultdict(list)
                     session = self._get_cached_session()
-                    query = session.query(Obj).filter(
-                        Obj.hashkey.in_(hashkeys)
-                    ).with_entities(Obj.pack_id, Obj.hashkey, Obj.offset, Obj.length, Obj.compressed,
-                                    Obj.size).order_by(Obj.offset)
-                    for res in query:
-                        packs[res[0]].append(ObjQueryResults(res[1], res[2], res[3], res[4], res[5]))
+                    for chunk in chunk_iterator(loose_not_found, size=self._IN_SQL_MAX_LENGTH):
+                        query = session.query(Obj).filter(
+                            Obj.hashkey.in_(chunk)
+                        ).with_entities(Obj.pack_id, Obj.hashkey, Obj.offset, Obj.length, Obj.compressed,
+                                        Obj.size).order_by(Obj.offset)
+                        for res in query:
+                            packs[res[0]].append(ObjQueryResults(res[1], res[2], res[3], res[4], res[5]))
 
-                    # I will construct here the really missing objects
-                    really_not_found = set(loose_not_found)
+                    # I will construct here the really missing objects.
+                    # I make a copy of the set.
+                    really_not_found = loose_not_found.copy()
 
                     for pack_int_id, pack_metadata in packs.items():
+                        # I remove those that I found
                         really_not_found.difference_update(obj.hashkey for obj in pack_metadata)
 
                         pack_path = self._get_pack_path_from_pack_id(str(pack_int_id))
@@ -768,12 +787,12 @@ class Container:
         # maintenance operation, so I don't have to worry about concurrency:
         # If I don't find it here, is should not appear midway during the rest of the process
 
-        # I check the hash keys that are already in the pack
-        existing_packed_hashkeys = session.query(Obj).filter(Obj.hashkey.in_(loose_objects)).with_entities(Obj.hashkey
-                                                                                                          ).all()
-        # The query returns a list of length-1 tuples,
-        # and we need to unpack them
-        existing_packed_hashkeys = [value for value, in existing_packed_hashkeys]
+        existing_packed_hashkeys = []
+
+        for chunk in chunk_iterator(loose_objects, size=self._IN_SQL_MAX_LENGTH):
+            # I check the hash keys that are already in the pack
+            for res in session.query(Obj).filter(Obj.hashkey.in_(chunk)).with_entities(Obj.hashkey).all():
+                existing_packed_hashkeys.append(res[0])
         # I remove them from the loose_objects list
         loose_objects.difference_update(existing_packed_hashkeys)
         # Now, I should be left only with objects with hash keys that are not yet known.

--- a/disk_objectstore/container.py
+++ b/disk_objectstore/container.py
@@ -852,7 +852,16 @@ class Container:
             # If we are here, things should be guaranteed by SQLite to be written to disk.
             # Then, it's safe to do some clean up loose objects, for each pack.
             for obj_hashkey in loose_objects_this_pack:
-                os.remove(self._get_loose_path_from_hashkey(obj_hashkey))
+                try:
+                    os.remove(self._get_loose_path_from_hashkey(obj_hashkey))
+                except PermissionError:
+                    # On Windows, I could get a "PermissionError: [WinError 32] The process cannot access the file
+                    # because it is begin used by another process: '<filename>'" if, while packing, some process is
+                    # reading the loose object.
+                    # In this case, I just ignore this and continue. The object is now packed, and there is still
+                    # the corresponding loose object, but this is not an error.
+                    # At the next packing operation the loose object will be inexpensively deleted.
+                    pass
 
     def add_streamed_objects_to_pack(self, stream_list, compress=False, open_streams=False):
         """Add objects directly to a pack, reading from a list of streams.

--- a/disk_objectstore/container.py
+++ b/disk_objectstore/container.py
@@ -328,7 +328,11 @@ class Container:
         :return: a byte stream with the object content.
         """
         with self.get_object_stream(hashkey) as handle:
-            return handle.read()
+            content = handle.read()
+            if not content and hashkey != 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855':
+                raise ValueError('empty data returned! {}, type={}'.format(hashkey, type(handle)))
+
+            return content
 
     @contextmanager
     def get_object_stream(self, hashkey):
@@ -819,14 +823,21 @@ class Container:
                     obj.length = pack_handle.tell() - obj.offset
                     session.add(obj)
 
-                # Let's commit and do final operations before closing the pack file.
-                # Committing as soon as we are done with one pack
-                # Note: because of the logic above, in theory this should not raise an IntegrityError!
-                session.commit()
+                # flush and sync to disk before closing
+                pack_handle.flush()
+                os.fsync(pack_handle.fileno())
 
-                # Clean up loose objects, for each pack
-                for obj_hashkey in loose_objects_this_pack:
-                    os.remove(self._get_loose_path_from_hashkey(obj_hashkey))
+            # OK, if we are here, file was flushed, synced to disk and closed.
+            # Let's commit then the information to the DB, so it's officially a
+            # packed object. Note: committing as soon as we are done with one pack,
+            # so if there's a problem with one pack we don't start operating on the next one
+            # Note: because of the logic above, in theory this should not raise an IntegrityError!
+            session.commit()
+
+            # If we are here, things should be guaranteed by SQLite to be written to disk.
+            # Then, it's safe to do some clean up loose objects, for each pack.
+            for obj_hashkey in loose_objects_this_pack:
+                os.remove(self._get_loose_path_from_hashkey(obj_hashkey))
 
     def add_streamed_objects_to_pack(self, stream_list, compress=False, open_streams=False):
         """Add objects directly to a pack, reading from a list of streams.
@@ -924,9 +935,16 @@ class Container:
                     # Append the new hash key to the list of hash keys to return
                     hashkeys.append(obj_dict['hashkey'])
 
-                # This pack file is complete.
-                # Let's commit before closing the pack file.
-                session.commit()
+                # flush and sync to disk before closing
+                pack_handle.flush()
+                os.fsync(pack_handle.fileno())
+
+            # OK, if we are here, file was flushed, synced to disk and closed.
+            # Let's commit then the information to the DB, so it's officially a
+            # packed object. Note: committing as soon as we are done with one pack,
+            # so if there's a problem with one pack we don't start operating on the next one
+            # Note: because of the logic above, in theory this should not raise an IntegrityError!
+            session.commit()
 
         return hashkeys
 

--- a/disk_objectstore/container.py
+++ b/disk_objectstore/container.py
@@ -342,11 +342,7 @@ class Container:
         :return: a byte stream with the object content.
         """
         with self.get_object_stream(hashkey) as handle:
-            content = handle.read()
-            if not content and hashkey != 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855':
-                raise ValueError('empty data returned! {}, type={}'.format(hashkey, type(handle)))
-
-            return content
+            return handle.read()
 
     @contextmanager
     def get_object_stream(self, hashkey):

--- a/disk_objectstore/container.py
+++ b/disk_objectstore/container.py
@@ -463,7 +463,9 @@ class Container:
                     obj_path = self._get_loose_path_from_hashkey(hashkey=loose_hashkey)
                     try:
                         last_open_file = open(obj_path, mode='rb')
-                        yield loose_hashkey, last_open_file, os.path.getsize(obj_path)
+                        # I do not use os.path.getsize in case the file has just
+                        # been deleted by a concurrent writer
+                        yield loose_hashkey, last_open_file, os.fstat(last_open_file.fileno()).st_size
                     except FileNotFoundError:
                         loose_not_found.add(loose_hashkey)
                         continue

--- a/disk_objectstore/examples/example_objectstore.py
+++ b/disk_objectstore/examples/example_objectstore.py
@@ -105,7 +105,9 @@ def main(num_files, min_size, max_size, directly_to_pack, path, clear, num_bulk_
 
         # Check that the content is correct
         for filename in retrieved:
-            assert retrieved[filename] == files[filename], 'Mismatch (content) for {}'.format(filename)
+            assert retrieved[filename] == files[filename], 'Mismatch (content) for {}, {} vs {}'.format(
+                filename, retrieved[filename], files[filename]
+            )
 
         # Check that num_files new loose files are present now
         counts = container.count_objects()
@@ -212,6 +214,7 @@ def main(num_files, min_size, max_size, directly_to_pack, path, clear, num_bulk_
     ########################################
     # THIRD: a lot of independent reads, one per object
     random.shuffle(random_keys)
+    retrieved = {}
     start = time.time()
     for filename in random_keys:
         obj_hashkey = hashkey_mapping[filename]
@@ -221,7 +224,9 @@ def main(num_files, min_size, max_size, directly_to_pack, path, clear, num_bulk_
     print('Time to retrieve {} packed objects in random order: {} s'.format(num_files, tot_time))
 
     for filename in retrieved:
-        assert retrieved[filename] == files[filename], 'Mismatch for {}'.format(filename)
+        assert retrieved[filename] == files[filename], 'Mismatch (content) for {}, {} vs {}'.format(
+            filename, retrieved[filename], files[filename]
+        )
 
     print('All tests passed')
 

--- a/disk_objectstore/utils.py
+++ b/disk_objectstore/utils.py
@@ -4,6 +4,7 @@ Some might be useful also for end users, like the wrappers to get streams,
 like the ``LazyOpener``.
 """
 import hashlib
+import itertools
 import os
 import uuid
 import zlib
@@ -543,3 +544,18 @@ class HashWriterWrapper:
     def fileno(self):
         """Return the integer file descriptor of the underlying file object."""
         return self._write_stream.fileno()
+
+
+def chunk_iterator(iterator, size):
+    """Given an iterator, split it in chunks of size `size` (except the last one, that can be shorter).
+
+    Examples:
+
+    - `list(chunk_iterator(range(10), 3))` returns `[(0, 1, 2), (3, 4, 5), (6, 7, 8), (9,)]`
+    - `list(chunk_iterator(range(9), 3))` returns `[(0, 1, 2), (3, 4, 5), (6, 7, 8)]`
+
+    :param iterator: an iterator to divide in chunks
+    :param size: the size of each chunk
+    """
+    iterator = iter(iterator)
+    return iter(lambda: tuple(itertools.islice(iterator, size)), ())

--- a/tests/concurrent_tests/periodic_worker.py
+++ b/tests/concurrent_tests/periodic_worker.py
@@ -8,14 +8,13 @@ same time, to check if there are issues while all of these work at the same time
 
 The script will also write the objects it created (in a safe way also when there is concurrency) in a folder
 (defined by the -s option). Then each locust will try to read back *all* files written by *all* locusts (including
-itself) and check if the MD5s are correct.
+itself) and check if the checksums are correct.
 """
 import datetime
 import hashlib
 import json
 import os
 import random
-import shutil
 import tempfile
 import time
 
@@ -47,7 +46,8 @@ def timestamp():
     '-s',
     '--shared-folder',
     default='/tmp/test-container-shared',
-    help='The path to a test folder in which all locusts will write the MD5s for others to read. It must already exist.'
+    help=
+    'The path to a test folder in which all locusts will write the checksums for others to read. It must already exist.'
 )
 @click.option('-b', '--bulk-read', is_flag=True, help='Whether to use bulk reads, or a loop on each hash key.')
 @click.help_option('-h', '--help')
@@ -57,6 +57,7 @@ def main(num_files, min_size, max_size, path, repetitions, wait_time, shared_fol
     if not os.path.isdir(shared_folder):
         raise ValueError("Create the folder '{}' first!".format(shared_folder))
     container = Container(path)
+    # In the tests we pass it already initialised, so this is never called
     if not container.is_initialised:
         print('Initialising the container...')
         container.init_container()
@@ -91,37 +92,45 @@ def main(num_files, min_size, max_size, path, repetitions, wait_time, shared_fol
         for content in contents:
             hashkeys.append(container.add_object(content))
 
-        md5s = {}
+        checksums = {}
         for obj_hashkey, content in zip(hashkeys, contents):
-            md5s[obj_hashkey] = hashlib.md5(content).hexdigest()
+            # I use the same sha256 algo, to check more easily which one is corrupt
+            checksum = hashlib.sha256(content).hexdigest()
+            try:
+                # If I already found the same hash key, check that also the checksum is the same
+                assert checksums[obj_hashkey] == checksum
+            except KeyError:
+                # Not found yet, it's OK
+                pass
+            checksums[obj_hashkey] = checksum
 
         ## Dump to file
         with tempfile.NamedTemporaryFile(mode='w', encoding='utf8', dir=shared_folder, delete=False) as fhandle:
             fname = fhandle.name
-            json.dump(md5s, fhandle)
+            json.dump(checksums, fhandle)
         # Atomic move in place (so other locusts don't try to read partially written files)
-        shutil.move(fname, '{}.md5'.format(fname))
+        os.rename(fname, '{}.sha'.format(fname))
 
         # Re-read everything, also from other processes
-        all_md5s = {}
+        all_checksums = {}
         file_count = 0
         for fname in os.listdir(shared_folder):
-            if not fname.endswith('.md5'):
+            if not fname.endswith('.sha'):
                 continue
             file_count += 1
             with open(os.path.join(shared_folder, fname)) as fhandle:
-                chunk_md5s = json.load(fhandle)
-            all_md5s.update(chunk_md5s)
+                chunk_checksums = json.load(fhandle)
+            all_checksums.update(chunk_checksums)
         print(
-            '[{} {}] {} object MD5s read from {} files ({}).'.format(
-                proc_id, timestamp(), len(all_md5s), file_count,
+            '[{} {}] {} object checksums read from {} files ({}).'.format(
+                proc_id, timestamp(), len(all_checksums), file_count,
                 'with bulk reads' if bulk_read else 'with single-object reads'
             )
         )
 
         ########################################
         # single bulk read
-        all_hashkeys = list(all_md5s.keys())
+        all_hashkeys = list(all_checksums.keys())
         random.shuffle(all_hashkeys)
 
         if bulk_read:
@@ -133,35 +142,40 @@ def main(num_files, min_size, max_size, path, repetitions, wait_time, shared_fol
                     retrieved_content[obj_hashkey] = container.get_object_content(obj_hashkey)
                 except NotExistent:
                     retrieved_content[obj_hashkey] = None
-        retrieved_md5s = {}
+        retrieved_checksums = {}
         for obj_hashkey, content in retrieved_content.items():
             if content is None:
                 raise ValueError('No content returned for object {}!'.format(obj_hashkey))
-            retrieved_md5s[obj_hashkey] = hashlib.md5(content).hexdigest()
+            if not content and obj_hashkey != 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855':
+                print('WARNING!!! {} is {} ({})'.format(obj_hashkey, content, type(content)))
+            retrieved_checksums[obj_hashkey] = hashlib.sha256(content).hexdigest()
 
-        only_left = set(retrieved_md5s).difference(all_md5s)
-        only_right = set(all_md5s).difference(retrieved_md5s)
-        assert not only_right, 'objects only in all_md5s: {}'.format(only_right)
-        assert not only_left, 'objects only in retrieved_md5s: {}'.format(only_left)
-        for key in retrieved_md5s:
-            assert all_md5s[key] == retrieved_md5s[key], 'Mismatch for {}: {} vs. {}'.format(
-                key, all_md5s[key], retrieved_md5s[key]
+        only_left = set(retrieved_checksums).difference(all_checksums)
+        only_right = set(all_checksums).difference(retrieved_checksums)
+        assert not only_right, 'objects only in all_checksums: {}'.format(only_right)
+        assert not only_left, 'objects only in retrieved_checksums: {}'.format(only_left)
+        for key in retrieved_checksums:
+            assert all_checksums[key] == retrieved_checksums[key], 'Mismatch for {}: {} vs. {}'.format(
+                key, all_checksums[key], retrieved_checksums[key]
             )
         del retrieved_content
 
         random.shuffle(all_hashkeys)
-        retrieved_md5s = {}
+        retrieved_checksums = {}
         for obj_hashkey in all_hashkeys:
-            retrieved_md5s[obj_hashkey] = hashlib.md5(container.get_object_content(obj_hashkey)).hexdigest()
+            content = container.get_object_content(obj_hashkey)
+            if not content and obj_hashkey != 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855':
+                print('WARNING!!! {} is {} ({})'.format(obj_hashkey, content, type(content)))
+            retrieved_checksums[obj_hashkey] = hashlib.sha256(content).hexdigest()
 
-        only_left = set(retrieved_md5s).difference(all_md5s)
-        only_right = set(all_md5s).difference(retrieved_md5s)
-        assert not only_right, 'objects only in all_md5s: {}'.format(only_right)
-        assert not only_left, 'objects only in retrieved_md5s: {}'.format(only_left)
-        for key in retrieved_md5s:
+        only_left = set(retrieved_checksums).difference(all_checksums)
+        only_right = set(all_checksums).difference(retrieved_checksums)
+        assert not only_right, 'objects only in all_checksums: {}'.format(only_right)
+        assert not only_left, 'objects only in retrieved_checksums: {}'.format(only_left)
+        for key in retrieved_checksums:
             try:
-                assert all_md5s[key] == retrieved_md5s[key], 'Mismatch for {}: {} vs. {}'.format(
-                    key, all_md5s[key], retrieved_md5s[key]
+                assert all_checksums[key] == retrieved_checksums[key], 'Mismatch for {}: {} vs. {}'.format(
+                    key, all_checksums[key], retrieved_checksums[key]
                 )
             except AssertionError:
                 loose_path = container._get_loose_path_from_hashkey(key)  # pylint: disable=protected-access

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -77,6 +77,11 @@ def test_deletion_while_open(temp_dir):
         # In either case (I got an exception on Windows, I could delete the file on POSIX)
         # I should still be able to read the correct content.
         # Notably, on POSIX I can read even if the file is not there anymore.
+
+        # I first check that I can get its size
+        assert os.fstat(fhandle.fileno()).st_size == len(content)
+
+        # And that I can read the content
         read_content = fhandle.read()
         assert read_content == content
 

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -1,0 +1,90 @@
+"""Test basic I/O functionality.
+
+This is also a way to verify the behavior of the underlying OS/filesystem.
+"""
+import os
+
+
+def test_concurrent_append_read(temp_dir):
+    """Check what happens when reading a file that is being written in append mode in the meantime."""
+    fname = os.path.join(temp_dir, 'test_file')
+
+    first_part = b'v92jgwkf'
+    intermediate = b'fbd9pjv2klmwg'
+    second_part = b'2flemawlefm'
+
+    with open(fname, 'ab') as write_handle:
+        write_handle.write(first_part)
+        # Flush to make sure everything is out of the buffer and visible to other processes
+        write_handle.flush()
+
+        # Verify that I can open in read ode and that I already find the content,
+        # even if the write_handle is not closed
+        with open(fname, 'rb') as read_handle:
+            # Read the first part, check it's correct
+            read_chunk = read_handle.read(len(first_part))
+            assert read_chunk == first_part
+
+            # Write the intermediate and the second part, and flush out buffers
+            write_handle.write(intermediate)
+            write_handle.write(second_part)
+            write_handle.flush()
+
+            # Check that the file size on disk is the expected one
+            assert os.path.getsize(fname) == len(first_part) + len(intermediate) + len(second_part)
+
+            # Now, jump to the beginning of the second part for reading (skipping 'intermediate')
+            read_handle.seek(len(first_part) + len(intermediate))
+            read_chunk = read_handle.read(len(second_part))
+            assert read_chunk == second_part
+
+
+def test_deletion_while_open(temp_dir):
+    """Check the different behavior of deletion of an open file (in read mode).
+
+    On POSIX: I can delete successfully the file and I can still read the content.
+    On Windows: I cannot delete a file while it is open.
+    """
+    fname = os.path.join(temp_dir, 'test_file')
+
+    content = b'rfr23ewv3wg4w'
+
+    # Write something to the file
+    with open(fname, 'wb') as fhandle:
+        fhandle.write(content)
+
+    # Now open again the file
+    with open(fname, 'rb') as fhandle:
+        # After opening, and before starting to read, I delete the file
+        try:
+            os.remove(fname)
+        except PermissionError as exc:
+            # On Windows, I should get:
+            # PermissionError: [WinError 32] The process cannot access the file because
+            # it is begin used by another process: '<filename>'
+            assert os.name == 'nt', 'I should get a PermissionError only on Windows!'
+
+            # Check the error code
+            # Notes:
+            # - errno.EACCES == 13
+            # - os.streerror(exc.errno) == 'Permission denied'
+            assert exc.errno == 13
+            assert os.path.isfile(fname), 'The file was actually deleted on Windows, unexpected!'
+        else:
+            assert os.name == 'posix', "I should be able to delete a file while it's still open only on POSIX!"
+            assert not os.path.isfile(fname), "The file wasn't really deleted in POSIX, unexpected!"
+
+        # In either case (I got an exception on Windows, I could delete the file on POSIX)
+        # I should still be able to read the correct content.
+        # Notably, on POSIX I can read even if the file is not there anymore.
+        read_content = fhandle.read()
+        assert read_content == content
+
+    # Now the file is closed. On POSIX, I know it's not there.
+    # Let me try to delete it also on Windows.
+    if os.name == 'nt':
+        # Now that the file is closed, I should be able to remove it
+        os.remove(fname)
+
+    # The file should not be there on any platform, now
+    assert not os.path.isfile(fname)

--- a/tests/test_concurrency.py
+++ b/tests/test_concurrency.py
@@ -3,6 +3,8 @@ import os
 import subprocess
 import sys
 
+import pytest
+
 from disk_objectstore import Container
 
 THIS_FILE_DIR = os.path.dirname(os.path.realpath(__file__))
@@ -11,6 +13,7 @@ CONCURRENT_DIR = os.path.join(THIS_FILE_DIR, 'concurrent_tests')
 NUM_WORKERS = 4
 
 
+@pytest.mark.xfail(os.name == 'nt', reason='Still some problems on Windows, see #4')
 def test_concurrency(temp_dir):
     """Test to run concurrently many workers creating objects, and at the same time one packer.
 

--- a/tests/test_concurrency.py
+++ b/tests/test_concurrency.py
@@ -3,8 +3,6 @@ import os
 import subprocess
 import sys
 
-import pytest
-
 from disk_objectstore import Container
 
 THIS_FILE_DIR = os.path.dirname(os.path.realpath(__file__))
@@ -13,7 +11,6 @@ CONCURRENT_DIR = os.path.join(THIS_FILE_DIR, 'concurrent_tests')
 NUM_WORKERS = 4
 
 
-@pytest.mark.xfail(reason='I know this is not yet working. This is tracked in #4')
 def test_concurrency(temp_dir):
     """Test to run concurrently many workers creating objects, and at the same time one packer.
 
@@ -51,12 +48,24 @@ def test_concurrency(temp_dir):
         )
 
     packer_out, packer_err = packer_proc.communicate()
+    if packer_out:
+        print('** STDOUT OF PACKER')
+        print(packer_out.decode('utf8'))
+    if packer_err:
+        print('** STDERR OF PACKER')
+        print(packer_err.decode('utf8'), file=sys.stderr)
     worker_outs = []
     worker_errs = []
-    for worker_proc in worker_procs:
+    for worker_id, worker_proc in enumerate(worker_procs):
         worker_out, worker_err = worker_proc.communicate()
         worker_outs.append(worker_out)
         worker_errs.append(worker_err)
+        if worker_out:
+            print('** STDOUT OF WORKER {}'.format(worker_id))
+            print(worker_out.decode('utf8'))
+        if worker_err:
+            print('** STDERR OF WORKER {}'.format(worker_id), file=sys.stderr)
+            print(worker_err.decode('utf8'), file=sys.stderr)
 
     error_messages = []
 

--- a/tests/test_concurrency.py
+++ b/tests/test_concurrency.py
@@ -32,14 +32,14 @@ def test_concurrency(temp_dir):
     os.mkdir(shared_dir)
 
     # Start the packer
-    packer_proc = subprocess.Popen([sys.executable, packer_script, '-p', container_dir, '-r', '5', '-w', '0.83'],
+    packer_proc = subprocess.Popen([sys.executable, packer_script, '-p', container_dir, '-r', '7', '-w', '0.83'],
                                    stdout=subprocess.PIPE,
                                    stderr=subprocess.PIPE)
 
     # Start the workers
     worker_procs = []
     for worker_id in range(NUM_WORKERS):
-        options = ['-r', '4', '-w', '0', '-p', container_dir, '-s', shared_dir]
+        options = ['-r', '6', '-w', '0', '-p', container_dir, '-s', shared_dir]
         if worker_id % 2:
             # One every two will read in bulk, the other within a loop
             options += ['-b']

--- a/tests/test_container.py
+++ b/tests/test_container.py
@@ -863,3 +863,122 @@ def test_get_objects_stream_closes(temp_container, generate_random_data):
             stream.read()
     # Check that at the end nothing is left open
     assert len(current_process.open_files()) == start_open_files
+
+
+def test_length_get_objects(temp_container):
+    """Check that the iterator to get the object streams does not perform unnecessary operations.
+
+    This is mostly to check for efficiency and that I don't iterate by mistake multiple times on the same object.
+    """
+    # Note: I created different data, so these 4 hashkeys will always be different
+    data = [b'1', b'2', b'3', b'4']
+    data_dict = {}
+    for val in data:
+        hashkey = temp_container.add_object(val)
+        data_dict[hashkey] = val
+
+    # Test 1: I pass all the known hashkeys: I iterate (in some order) on all of them
+    hashkeys = list(data_dict)
+    iterated_hashkeys = []
+    with temp_container.get_objects_stream_and_size(hashkeys=hashkeys) as triplets:
+        for obj_hashkey, _, _ in triplets:
+            iterated_hashkeys.append(obj_hashkey)
+    assert sorted(iterated_hashkeys) == sorted(hashkeys)
+
+    # Test 2: I pass half the known hashkeys: I iterate (in some order) on all of them
+    hashkeys = list(data_dict)[:2]
+    iterated_hashkeys = []
+    with temp_container.get_objects_stream_and_size(hashkeys=hashkeys) as triplets:
+        for obj_hashkey, _, _ in triplets:
+            iterated_hashkeys.append(obj_hashkey)
+    assert sorted(iterated_hashkeys) == sorted(hashkeys)
+
+    # Test 3: I pass half the known hashkeys, twice: I iterate (in some order) on all of them, BUT ONLY ONCE
+    partial_hashkeys = list(data_dict)[:2]
+    hashkeys = partial_hashkeys + partial_hashkeys
+    iterated_hashkeys = []
+    with temp_container.get_objects_stream_and_size(hashkeys=hashkeys) as triplets:
+        for obj_hashkey, _, _ in triplets:
+            iterated_hashkeys.append(obj_hashkey)
+    # I should iterate on them ONLY ONCE
+    assert sorted(iterated_hashkeys) == sorted(partial_hashkeys)
+
+    # Test 4A: I pass half the known hashkeys, twice, PLUS SOME UNKNOWN HASHKEY, also twice
+    # If skip_if_missing=False, I should iterate on ALL of them, BUT ONLY ONCE
+    unknown_hashkeys = ['unk1', 'unk2']
+    partial_hashkeys = list(data_dict)[:2]
+    hashkeys = partial_hashkeys + partial_hashkeys + unknown_hashkeys + unknown_hashkeys
+    iterated_hashkeys = []
+    with temp_container.get_objects_stream_and_size(hashkeys=hashkeys, skip_if_missing=False) as triplets:
+        for obj_hashkey, _, _ in triplets:
+            iterated_hashkeys.append(obj_hashkey)
+    # I should iterate on them ONLY ONCE
+    assert sorted(iterated_hashkeys) == sorted(set(hashkeys))
+
+    # Test 4B: I pass half the known hashkeys, twice, PLUS SOME UNKNOWN HASHKEY, also twice
+    # If skip_if_missing=True, I should iterate ONLY on the known one, BUT ONLY ONCE
+    unknown_hashkeys = ['unk1', 'unk2']
+    partial_hashkeys = list(data_dict)[:2]
+    hashkeys = partial_hashkeys + partial_hashkeys + unknown_hashkeys + unknown_hashkeys
+    iterated_hashkeys = []
+    with temp_container.get_objects_stream_and_size(hashkeys=hashkeys, skip_if_missing=True) as triplets:
+        for obj_hashkey, _, _ in triplets:
+            iterated_hashkeys.append(obj_hashkey)
+    # I should iterate on them ONLY ONCE
+    assert sorted(iterated_hashkeys) == sorted(set(partial_hashkeys))
+
+
+def test_very_long_hashkeys_list(temp_container):
+    """Test that even when asking for *a lot* of hashkeys in one shot, this does not fail.
+
+    Indeed, in SQLite, there is a maximum number of paramaters in a given query. And SQLAlchemy
+    converts `.in_()` filters to parameters.
+    In the code, we chunk any request to make sure it does not crash because we're asking more
+    entries than SQLite can support (the limit unfortunately is decided at compile-time (of SQLite)
+    and so it's not under our control).
+    """
+    # A bit more than 10 times the internal maximum length of SQL queries
+    # This is ~9500, and should exceed the hardcoded limits of 999 on at least some platforms
+    # I put +3 to have something that is NOT a multiple
+    # Note: this is already a relatively large number and the test takes a few seconds
+    # (because I'm writing loose objects as well)
+    num_objects = temp_container._IN_SQL_MAX_LENGTH * 10 + 3  # pylint: disable=protected-access
+    # Generate a long list of objects with *different* data, so they are not deduplicated
+    data = [str(i).encode('ascii') for i in range(num_objects)]
+
+    # The container is empty: let's check that asking for unknown objects work
+    # NOTE: I just generate some random 'unknown' hash key
+    res = temp_container.get_objects_content(['unk{}'.format(i) for i in range(num_objects)])
+    assert not res
+
+    # I now add all objects
+    hashkeys = []
+    for val in data:
+        hashkeys.append(temp_container.add_object(val))
+
+    # This should always be true! I am just appending different objects
+    assert len(hashkeys) == num_objects
+
+    # Let's ask for all of them; this should work even if they are many
+    values = temp_container.get_objects_content(hashkeys)
+    assert values == dict(zip(hashkeys, data))
+
+    # Let's pack everything. This should work even if there are many objects
+    # (also in here there are .in_ calls)
+    temp_container.pack_all_loose()
+
+    # Let's ask back again all data, now that it's packed
+    # Also for packed objects, asking for a lot of data should be OK
+    values = temp_container.get_objects_content(hashkeys)
+    assert values == dict(zip(hashkeys, data))
+
+    # Create more data, different than the one before and add directly to packs, this time.
+    # This should also work
+    direct_pack_data = ['P{}'.format(i).encode('ascii') for i in range(num_objects)]
+    direct_pack_hashkeys = temp_container.add_objects_to_pack(direct_pack_data)
+
+    # Finally, ask back everything and check
+    expected_values = dict(zip(hashkeys, data))
+    expected_values.update(dict(zip(direct_pack_hashkeys, direct_pack_data)))
+    values = temp_container.get_objects_content(hashkeys + direct_pack_hashkeys)
+    assert values == expected_values

--- a/tests/test_container.py
+++ b/tests/test_container.py
@@ -1,4 +1,5 @@
 """Test of the object-store container module."""
+# pylint: disable=too-many-lines
 import hashlib
 import io
 import os
@@ -6,6 +7,7 @@ import random
 import shutil
 import tempfile
 import zlib
+import pathlib
 
 import psutil
 import pytest
@@ -982,3 +984,81 @@ def test_very_long_hashkeys_list(temp_container):
     expected_values.update(dict(zip(direct_pack_hashkeys, direct_pack_data)))
     values = temp_container.get_objects_content(hashkeys + direct_pack_hashkeys)
     assert values == expected_values
+
+
+@pytest.mark.parametrize('compress', [True, False])
+def test_simulate_concurrent_packing(temp_container, compress):  # pylint: disable=invalid-name
+    """Simulate race conditions while reading and packing."""
+    content = b'abc'
+    hashkey = temp_container.add_object(content)
+
+    loose_dir_path = pathlib.Path(temp_container._get_loose_folder())  # pylint: disable=protected-access
+    with temp_container.get_object_stream(hashkey) as fhandle:
+        fname = fhandle.name
+        # Check that this is a loose object (i.e. that it is in the loose folder)
+        assert loose_dir_path in pathlib.Path(fname).parents
+        assert fhandle.read(1) == b'a'
+        temp_container.pack_all_loose(compress=compress)
+        assert fhandle.read() == b'bc'
+
+    # On Windows, the loose file might still be there because I cannot delete an open file
+    # Still, this is a valid situation and I should be able to get the correct content anyway
+    # The following line should be read from the pack file. Anyway, the important thing to test
+    # is that the content is properly returned
+    with temp_container.get_object_stream(hashkey) as fhandle:
+        assert fhandle.read() == b'abc'
+
+    temp_container.pack_all_loose(compress=compress)
+    # After a second packing, the loose file *must* have been removed
+    assert not os.path.exists(fname)
+
+
+@pytest.mark.parametrize('compress', [True, False])
+def test_simulate_concurrent_packing_multiple(temp_container, compress):  # pylint: disable=invalid-name
+    """Simulate race conditions while reading and packing."""
+    content1 = b'abc'
+    content2 = b'def'
+    hashkey1 = temp_container.add_object(content1)
+    hashkey2 = temp_container.add_object(content2)
+    data = {hashkey1: content1, hashkey2: content2}
+
+    loose_dir_path = pathlib.Path(temp_container._get_loose_folder())  # pylint: disable=protected-access
+    with temp_container.get_objects_stream_and_size([hashkey1, hashkey2]) as triplets:
+        counter = 0
+
+        for obj_hashkey, stream, size in triplets:
+            if counter == 0:
+                # In the first step, I always pack the rest
+                fname = stream.name
+                # Check that this is a loose object (i.e. that it is in the loose folder)
+                assert loose_dir_path in pathlib.Path(fname).parents
+                if obj_hashkey == hashkey1:
+                    assert stream.read(1) == b'a'
+                    temp_container.pack_all_loose(compress=compress)
+                    assert stream.read() == b'bc'
+                elif obj_hashkey == hashkey2:
+                    assert stream.read(1) == b'd'
+                    temp_container.pack_all_loose(compress=compress)
+                    assert stream.read() == b'ef'
+                else:
+                    # Should not happen!
+                    raise ValueError('Unknown hash key {}'.format(obj_hashkey))
+            elif counter == 1:
+                # This should be the other object
+                # This was loose when get_objects_stream_and_size was called, but was packed in the meantime
+                # I should still be able to get it correctly
+                assert data[obj_hashkey] == stream.read()
+            else:
+                # Should not happen!
+                raise ValueError('There should be only two objects!')
+            counter += 1
+
+    # On Windows, the loose file might still be there because I cannot delete an open file
+    # Still, this is a valid situation and I should be able to get the correct content anyway
+    # The following line should be read from the pack file. Anyway, the important thing to test
+    # is that the content is properly returned
+    assert data == temp_container.get_objects_content([hashkey1, hashkey2])
+
+    temp_container.pack_all_loose(compress=compress)
+    # After a second packing, the loose file *must* have been removed
+    assert not os.path.exists(fname)

--- a/tests/test_utils.py
+++ b/tests/test_utils.py
@@ -597,3 +597,16 @@ def test_hash_writer_wrapper(temp_dir, hash_type, expected_hash):
 
     with open(os.path.join(temp_dir, filename), 'rb') as fhandle:
         assert fhandle.read() == content
+
+
+def test_chunk_iterator():
+    """Test the correct functionality of the chunk iterator."""
+    # Using `iter(range())` that is an iterator and does not have a length or allows
+    # getting an element by index (i.e., `iter(range(10))` does not work), to make
+    # sure this works also for iterators and not only for lists or similar
+
+    # Check for lengths exactly multiple of the size
+    assert list(utils.chunk_iterator(iter(range(9)), 3)) == [(0, 1, 2), (3, 4, 5), (6, 7, 8)]
+
+    # Check for lengths that give a remainder
+    assert list(utils.chunk_iterator(iter(range(10)), 3)) == [(0, 1, 2), (3, 4, 5), (6, 7, 8), (9,)]

